pkg/descheduler/node/node.go

```diff
@@ -111,3 +112,17 @@ func IsReady(node *v1.Node) bool {
 	}*/
 	return true
 }
+
+// IsNodeUschedulable checks if the node is unschedulable. This is helper function to check only in case of
+// underutilized node so that they won't be accounted for.
+func IsNodeUschedulable(node *v1.Node) bool {
+	if node.Spec.Unschedulable {
+		return true
+	}
+	for _, taint := range node.Spec.Taints {
+		if taint.Effect == v1.TaintEffectNoSchedule {
```

Comment:
sorry, just noticed this. This is not true as some pods that tolerate that NoSchedule taint can still be scheduled, or unless we find a way in upstream to make node unschedulable using some standard taint where no pod could tolerate that taint. Its better to remove this.

Reply:
You are right when it comes to NoSchedule taint but this is what node controller and kubeadm are using. I believe the idea is to slowly move away from Unschedulable field in nodeSpec. So, to be in sync with upstream community, I think it would be better to respect `NoSchedule` taint but add a TODO saying that its not proper UnSchedulable field. WDYT?

Reply:
No that does not seem correct. Not sure what you mean by " this is what node controller and kubeadm are using. "? Are they using some taint to declare a node as Unschedulable? if yes, could you point me to that? As I said, without having a NoSchedule taint that "no pod" can tolerate, we can not use NoSchedule taint as an indicator for Unschedulable. Also  not sure what do you mean by " to be in sync with upstream community"? I really want to be in sync, but only if there is a solution upstream.

Reply:
I agree with you that it is not the correct way to do it and I believe you created an issue upstream for it. But kubeadm moved to taint for master(https://github.com/kubernetes/kubernetes/pull/41835/files). I just created a cluster using kubeadm and the master just has label and a NoSchedule taint. No Unschedulable field, if we go with our current approach, we will take nodes with NoSchedule taint into account as well. 

```
Name:			descheduler-54486c3a-c257-11e7-a71e-c85b76333877
Role:			
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/os=linux
			kubernetes.io/hostname=descheduler-54486c3a-c257-11e7-a71e-c85b76333877
			node-role.kubernetes.io/master=
Annotations:		node.alpha.kubernetes.io/ttl=0
			volumes.kubernetes.io/controller-managed-attach-detach=true
Taints:			node-role.kubernetes.io/master:NoSchedule
CreationTimestamp:	Sun, 05 Nov 2017 13:33:00 -0500
Conditions:
```

I am fine with either with/without NoSchedule Taint.

Reply:
you also need to check if they are not allowing pods to set "node-role.kubernetes.io/master:NoSchedule" via admission controller or some other means. Anyway we can *NOT* consider *any* NoSchedule taint for Unschedulable something this PR does so its better to remove that check.